jupyterhub/tests/selenium/test_browser.py

```diff
@@ -854,3 +857,148 @@ async def test_user_logout(app, browser, url, user):
+    assert expected_client_id, expected_redirect_url in browser.current_url
+
+    # login user
+    await login(browser, user.name, pass_w=str(user.name))
+    auth_button = browser.find_element(By.XPATH, '//input[@type="submit"]')
+    if not auth_button.is_displayed():
+        await webdriver_wait(
+            browser,
+            EC.visibility_of_element_located((By.XPATH, '//input[@type="submit"]')),
+        )
+    # verify that user can see the service name and oauth URL
+    text_permission = browser.find_element(
+        By.XPATH, './/h1[text()="Authorize access"]//following::p'
+    ).text
+    assert f"JupyterHub service {service.name}", (
+        f"oauth URL: {expected_redirect_url}" in text_permission
+    )
+    # permissions check
+    oauth_form = browser.find_element(By.TAG_NAME, "form")
+    scopes_elements = oauth_form.find_elements(
+        By.XPATH, '//input[@type="hidden" and @name="scopes"]'
+    )
+    scope_list_oauth_page = []
+    for scopes_element in scopes_elements:
+        # checking that scopes are invisible on the page
+        assert not scopes_element.is_displayed()
+        scope_value = scopes_element.get_attribute("value")
+        scope_list_oauth_page.append(scope_value)
+
+    # checking that all scopes granded to user are presented in POST form (scope_list)
+    assert all(x in scope_list_oauth_page for x in user_scopes)
+    assert f"access:services!service={service.name}" in scope_list_oauth_page
+
+    check_boxes = oauth_form.find_elements(
+        By.XPATH, '//input[@type="checkbox" and @name="raw-scopes"]'
+    )
+    for check_box in check_boxes:
+        # checking that user cannot uncheck the checkbox
+        assert not check_box.is_enabled()
+        assert check_box.get_attribute("disabled")
+        assert check_box.get_attribute("title") == "This authorization is required"
+
+    # checking that appropriete descriptions are displayed depending of scopes
+    descriptions = oauth_form.find_elements(By.TAG_NAME, 'span')
+    desc_list_UI = [
+        description.get_attribute("innerHTML")
+        .strip()
+        .replace('  ', '')
+        .replace('\n', '')
+        for description in descriptions
+    ]
+    # getting descriptions from scopes.py to compare them with descriptions on UI
+    scope_def = scopes.describe_raw_scopes(user_scopes, user.name)
+    list_of_desc_for_comparing = []
+    for scope_def in scope_def:
+        if "description" in scope_def:
+            text_desc_scope = scope_def.get("description")
+            text_filter = scope_def.get("filter")
+            if text_filter == "":
+                list_of_desc_for_comparing.append(text_desc_scope)
+            elif text_filter != "":
+                list_of_desc_for_comparing.append(
+                    text_desc_scope + f"Applies to {text_filter}."
+                )
+    assert sorted(desc_list_UI) == sorted(list_of_desc_for_comparing)
```

Comment:
We can be a little simpler here, and clearer with variable names:

```suggestion
    scope_descriptions = scopes.describe_raw_scopes(
        user_scopes or ['(no_scope)'], user.name
    )
    desc_list_expected = []
    for scope_description in scope_descriptions:
        description = scope_description.get("description")
        text_filter = scope_description.get("filter")
        if text_filter:
            description = f"{description} Applies to {text_filter}."
        desc_list_expected.append(description)

    assert sorted(desc_list_form) == sorted(desc_list_expected)
```